toolchain/parse/handle_choice.cpp

```diff
@@ -0,0 +1,97 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/parse/context.h"
+
+namespace Carbon::Parse {
+auto HandleChoiceIntroducer(Context& context) -> void {
+  auto state = context.PopState();
+
+  context.PushState(state, State::ChoiceDefinitionStart);
+  context.PushState(State::DeclNameAndParamsAsOptional, state.token);
+}
+
+auto HandleChoiceDefinitionStart(Context& context) -> void {
+  auto state = context.PopState();
+
+  if (!context.PositionIs(Lex::TokenKind::OpenCurlyBrace)) {
+    if (!state.has_error) {
+      CARBON_DIAGNOSTIC(ExpectedChoiceDefinition, Error,
+                        "Choice definition expected.");
+      context.emitter().Emit(*context.position(), ExpectedChoiceDefinition);
+    }
+
+    context.AddNode(NodeKind::ChoiceDefinitionStart, *context.position(),
+                    state.subtree_start, /*has_error=*/true);
+
+    context.AddNode(NodeKind::ChoiceDefinition, *context.position(),
+                    state.subtree_start, /*has_error=*/true);
+
+    context.SkipPastLikelyEnd(*context.position());
+    return;
+  }
+
+  context.AddNode(NodeKind::ChoiceDefinitionStart, context.Consume(),
+                  state.subtree_start, state.has_error);
+
+  state.has_error = false;
+  state.state = State::ChoiceDefinitionFinish;
+  context.PushState(state);
+
+  if (!context.PositionIs(Lex::TokenKind::CloseCurlyBrace)) {
+    context.PushState(State::ChoiceAlternative);
+  }
+}
+
+auto HandleChoiceAlternative(Context& context) -> void {
+  auto state = context.PopState();
+  if (!context.PositionIs(Lex::TokenKind::Identifier)) {
+    if (!state.has_error) {
+      CARBON_DIAGNOSTIC(ExpectedChoiceAlternativeName, Error,
+                        "Expected choice alternative name.");
+      context.emitter().Emit(*context.position(),
+                             ExpectedChoiceAlternativeName);
+    }
+
+    context.SkipPastLikelyEnd(*context.position());
+
+    state.state = State::ChoiceAlternativeFinish;
+    state.subtree_start = {};
+    state.token = *context.position();
+    state.has_error = true;
+    context.PushState(state);
```

Comment:
I think we're not getting any benefit from reusing the previous state here, given that we overwrite all of its fields. I think it'd be a little more obvious what we're doing if we push a fresh state:
```suggestion
    context.PushState(State::ChoiceAlternativeFinish);
    context.ReturnErrorOnState();
```
(It might also be reasonable to move the `PushState(ChoiceAlternativeFinish)` calls from here and line 68 to before the `if (!context.PositionIs...`, given that we do that unconditionally.)